kubectl exec/attach using websocket

robberphex opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Implements #89163

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

robberphex commented:
I signed it

robberphex commented:
I signed it

aojea commented:
current commit history is not easy to review, it seems it maintains the order used for local development but not for code review.

Also, the BUILD files are no longer needed

neolit123 commented:
/remove-area kubeadm
/remove-sig cluster-lifecycle

robberphex committed:
kubectl exec/attach using websocket

Co-authored-by: mlbiam <marc@tremolo.io>

robberphex commented:
> current commit history is not easy to review, it seems it maintains the order used for local development but not for code review.
> 
> Also, the BUILD files are no longer needed

I was hoping to keep commits from `mlbiam <marc@tremolo.io>`, so the commit log is not good to review. Fixed this, and removed useless BUILD file.

haosdent commented:
/remove-sig scheduling

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/websocket/connection.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// CreateStream creates a stream by the channel id
+func (connection *Connection) CreateStrean(channelID int) (httpstream.Stream, error) {
```
is this a typo? I can't find it in other place

aojea commented:
> I was hoping to keep commits from `mlbiam <marc@tremolo.io>`, so the commit log is not good to review. Fixed this, and removed useless BUILD file.

sorry, I didn't mean to squash all, but to have commits that were reviewable, previously they were containing comments and code that was being discarded between commits

fedebongio commented:
/assign @caesarxuchao 
/triage accepted

fedebongio commented:
/cc @zhang-xuebin

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go:
```diff
@@ -0,0 +1,479 @@
+			defer once.Do(func() { p.remoteStdinIn.Close() })
+
+			// this "copy" doesn't actually read anything - it's just here to wait for
+			// the server to close remoteStdin.
```
Isn't `remosteStdinIn` part of a local pipe? Why would server close it?

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go:
```diff
@@ -0,0 +1,479 @@
+	doneChan := make(chan struct{}, 2)
+
+	// set up error stream
+	p.errorStreamIn, p.errorStreamOut = io.Pipe()
```
Why do we need these pipes? Can we stitch together the websocket connection with the streamOption's stdin/stdout/stderr directly?

caesarxuchao reviewed (commented):
Thanks for reviving this, @robberphex.

Some high-level thoughts:
* We implemented a dumb httpstream.Connection interface which required a lot of boilerplate. But `httpstream.Connection` is SPDY specific, which supports multiplexing "streams" on the same "connection". WebSocket doesn't have a native multiplexing support. We should consider giving up the `httpstream.Connection` interface when implementing WebSocket.

* the remotecommandwebsocket/v4.go needs some clean up. I'm not sure why we need to create a set of intermediary pipes there. Also we should consider if we can deduplicate remotecommandwebsocket and the existing remotecommand package.

* We need to incorporate the latest executor/spdy improvements (#105632, #103177, and perhaps #107605) into the websocket implementation.

I haven't looked at the unit tests yet.

caesarxuchao commented:
cc @dberkov

jiahuif commented:
/cc

caesarxuchao commented:
/assign @jpbetz @yliaog 

in case I don't have bandwidth to review this PR in the future.

enj commented:
There is ongoing [discussion](https://groups.google.com/g/kubernetes-sig-api-machinery/c/KKAEZMIJn5U/m/gB1fQLmbAgAJ) around this topic and it is clear that such a change will require a KEP to move forward.  @robberphex please collaborate with @ash2k on next steps.

/close
